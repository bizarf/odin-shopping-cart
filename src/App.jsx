import React, { useState } from "react";
import Router from "./Router";

const App = () => {
    const [cart, setCart] = useState([]);

    const addToCart = (productId, productName, image, price) => {
        const cartIndex = cart.findIndex((id) => {
            return id.productId === productId;
        });
        // check if the product is already in the cart to avoid adding the product twice. if it's already in the cart, then increase the quantity by 1
        if (cartIndex === -1) {
            const item = {
                productId: productId,
                productName: productName,
                image: image,
                price: price,
                quantity: 1,
            };
            setCart([...cart, item]);
        } else {
            increaseQuantity(productId);
        }
    };

    const removeItemFromCart = (productId) => {
        const cartIndex = cart.findIndex((id) => {
            return id.productId === productId;
        });
        // make a copy of the array as we don't want to directly mutate it
        const updatedCart = cart;
        // remove the item with the index that was found in the above lines
        updatedCart.splice(cartIndex, 1);
        // update the cart's state by wrapping it in square brackets, adding a spread operator and the updated cart array
        setCart([...updatedCart]);
    };

    const increaseQuantity = (productId) => {
        const cartIndex = cart.findIndex((id) => {
            return id.productId === productId;
        });
        const updatedCart = cart;
        updatedCart[cartIndex].quantity++;
        setCart([...updatedCart]);
    };

    const decreaseQuantity = (productId) => {
        const cartIndex = cart.findIndex((id) => {
            return id.productId === productId;
        });
        const updatedCart = cart;
        if (updatedCart[cartIndex].quantity === 1) {
            removeItemFromCart(productId);
        } else {
            updatedCart[cartIndex].quantity--;
            setCart([...updatedCart]);
        }
    };

    return (
        <div className="app">
            <Router
                cart={cart}
                setCart={setCart}
                addToCart={addToCart}
                removeItemFromCart={removeItemFromCart}
                increaseQuantity={increaseQuantity}
                decreaseQuantity={decreaseQuantity}
            />
        </div>
    );
};

export default App;
