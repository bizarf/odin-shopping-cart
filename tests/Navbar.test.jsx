import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../src/App";
import { describe, expect, it } from "vitest";

describe("Navbar tests", () => {
    it("The Navbar renders", () => {
        render(<App />);
        const homeLink = screen.getByText("Home");
        const shopLink = screen.getByText("Shop");
        const cartLink = screen.getByText("Shopping Cart (0)");
        expect(homeLink).toMatchSnapshot();
        expect(shopLink).toMatchSnapshot();
        expect(cartLink).toMatchSnapshot();
    });
});

describe("Router tests", () => {
    it("The app goes from the homepage to the shop page", async () => {
        const user = userEvent.setup();
        render(<App />);
        const shopLink = screen.getByRole("link", { name: "Shop" });
        await user.click(shopLink);
        const shopHeader = screen.getByRole("heading", { name: "Shop" });
        expect(shopHeader).toBeInTheDocument();
    });

    it("The app goes from the homepage to the cart page", async () => {
        const user = userEvent.setup();
        render(<App />);
        const cartLink = screen.getByRole("link", {
            name: "Shopping Cart (0)",
        });
        await user.click(cartLink);
        const cartHeader = screen.getByRole("heading", {
            name: "My Shopping Cart",
        });
        expect(cartHeader).toBeInTheDocument();
    });

    it("The app goes back to the homepage from the shop page", async () => {
        const user = userEvent.setup();
        render(<App />);
        const homeLink = screen.getByRole("link", { name: "Home" });
        const shopLink = screen.getByRole("link", { name: "Shop" });
        await user.click(shopLink);
        await user.click(homeLink);
        const homePageText = screen.getByText("Welcome to Computer World!");
        expect(homePageText).toBeInTheDocument();
    });
});
