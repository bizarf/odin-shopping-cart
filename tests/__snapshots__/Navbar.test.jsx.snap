// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar tests > The Navbar renders 1`] = `
<a
  class="navLinks"
  href="#/"
>
  Home
</a>
`;

exports[`Navbar tests > The Navbar renders 2`] = `
<a
  class="navLinks"
  href="#/shop"
>
  Shop
</a>
`;

exports[`Navbar tests > The Navbar renders 3`] = `
<a
  class="navLinks"
  href="#/cart"
>
  Shopping Cart (
  0
  )
</a>
`;
